ocsp.go

```diff
@@ -140,117 +168,209 @@ func retryRevocationStatusCheck(totalTimeout *time.Duration, sleepTime time.Dura
+			code: ocspFailedDecomposeRequest,
 			err:  err,
 		}
 	}
 
 	// encode CertID, used as a key in the cache
-	hashAlg := getOIDFromHashAlgorithm(r.HashAlgorithm)
+	encodedCertID := &certIDKey{
+		r.HashAlgorithm,
+		base64.StdEncoding.EncodeToString(r.IssuerNameHash),
+		base64.StdEncoding.EncodeToString(r.IssuerKeyHash),
+		r.SerialNumber.String(),
+	}
+	return encodedCertID, &ocspStatus{
+		code: ocspSuccess,
+	}
+}
+
+func encodeCertIDKey(certIDKeyBase64 string) *certIDKey {
+	r, err := base64.StdEncoding.DecodeString(certIDKeyBase64)
+	if err != nil {
+		return nil
+	}
+	var c certID
+	rest, err := asn1.Unmarshal(r, &c)
+	if err != nil {
+		// error in parsing
+		return nil
+	}
+	if len(rest) > 0 {
+		// extra bytes to the end
+		return nil
+	}
+	return &certIDKey{
+		getHashAlgorithmFromOID(c.HashAlgorithm),
+		base64.StdEncoding.EncodeToString(c.NameHash),
+		base64.StdEncoding.EncodeToString(c.IssuerKeyHash),
+		c.SerialNumber.String(),
+	}
+}
+
+func decodeCertIDKey(k *certIDKey) string {
+	serialNumber := new(big.Int)
+	serialNumber.SetString(k.SerialNumber, 10)
+	nameHash, err := base64.StdEncoding.DecodeString(k.NameHash)
+	if err != nil {
+		return ""
+	}
+	issuerKeyHash, err := base64.StdEncoding.DecodeString(k.IssuerKeyHash)
+	if err != nil {
+		return ""
+	}
 	encodedCertID, err := asn1.Marshal(certID{
 		pkix.AlgorithmIdentifier{
-			Algorithm:  hashAlg,
+			Algorithm:  getOIDFromHashAlgorithm(k.HashAlgorithm),
 			Parameters: asn1.RawValue{Tag: 5 /* ASN.1 NULL */},
 		},
-		r.IssuerNameHash,
-		r.IssuerKeyHash,
-		r.SerialNumber,
+		nameHash,
+		issuerKeyHash,
+		serialNumber,
 	})
 	if err != nil {
-		return nil, &ocspStatus{
-			code: ocspFailedEncodeCertID,
-			err:  err,
-		}
-	}
-	return encodedCertID, &ocspStatus{
-		code: ocspSuccess,
+		return ""
 	}
+	return base64.StdEncoding.EncodeToString(encodedCertID)
 }
 
-func checkOCSPResponseCache(encodedCertID []byte, subject, issuer *x509.Certificate) *ocspStatus {
-	encodedCertIDBase64 := base64.StdEncoding.EncodeToString(encodedCertID)
-	ocspResponseCacheLock.Lock()
-	gotValueFromCache := ocspResponseCache[encodedCertIDBase64]
-	ocspResponseCacheLock.Unlock()
+func checkOCSPResponseCache(encodedCertID *certIDKey, subject, issuer *x509.Certificate) *ocspStatus {
+	ocspResponseCacheLock.RLock()
+	gotValueFromCache := ocspResponseCache[*encodedCertID]
+	ocspResponseCacheLock.RUnlock()
 	if len(gotValueFromCache) != 2 {
 		return &ocspStatus{
 			code: ocspMissedCache,
-			err:  fmt.Errorf("miss cache data. CertID: %v", encodedCertIDBase64),
+			err:  fmt.Errorf("miss cache data. CertID: %v", encodedCertID),
 		}
 	}
-	glog.V(2).Infof("hit cache: %v, time: %v, value: %v", encodedCertIDBase64, gotValueFromCache[0], gotValueFromCache[1])
+	glog.V(2).Infof("hit cache: %v, time: %v, value: %v", encodedCertID, gotValueFromCache[0], gotValueFromCache[1])
 	currentTime := float64(time.Now().UTC().Unix())
 	if epoch, ok := gotValueFromCache[0].(float64); ok {
```

Comment:
Can we change the name from epoch to something else? Like timestamp/ ts? Epoch is confusing as it usually would mean the Unix epoch time.

Reply:
ok. it is unix epoch time, though, ts is fine to me.